File: String/Determine_two_string_are_anagram_or_not/isAnagram.py
# Function to check if X and Y are anagrams or not
def isAnagram(X, Y):

    # if X's length is not same as Y's, they can't be anagram
    if len(X) != len(Y):
        return False

    # Create an empty dictionary
    freq = {}

    # Maintain count of each character of X in the dict
    for i in range(len(X)):
        freq[X[i]] = freq.get(X[i], 0) + 1

    # Do for each character of Y
    for i in range(len(Y)):

        # if y is not found in dict i. e. either y is not present
        # in X or has more occurrences in Y
        if not Y[i] in freq:
            return False


        # decrease the frequency of y in the dict
        freq[Y[i]] = freq[Y[i]] - 1

        # if its frequency becomes 0, erase it from dict
        if freq[Y[i]] == 0:
            del freq[Y[i]]

    # return true if dict becomes empty
    return not freq


if __name__ == "__main__":

    X = "listen"
    Y = "silent"

    if isAnagram(X, Y):
        print("Anagram")
    else:
        print("Not an Anagram")

File: String/Determine_two_string_are_anagram_or_not/IsAnagram.java
import java.util.HashMap;
import java.util.Map;

class IsAnagram {
	// Function to check if X and Y are anagrams or not
	public static boolean isAnagram(String X, String Y) {
		// if X's length is not same as Y's, they can't be anagram
		if (X.length() != Y.length()) {
			return false;
		}

		// create an empty map
		Map<Character, Integer> freq = new HashMap<>();

		// maintain count of each character of X in the map
		for (char c: X.toCharArray()) {
			freq.put(c, freq.getOrDefault(c, 0) + 1);
		}

		// do for each character of Y
		for (char c: Y.toCharArray()) {
			// if y is not found in map i.e. either y is not present
			// in String X or has more occurrences in String Y
			if (!freq.containsKey(c)) {
				return false;
			}

			// decrease the frequency of y in the map
			freq.put(c, freq.get(c) - 1);

			// if its frequency become 0, erase it from map
			if (freq.get(c) == 0) {
				freq.remove(c);
			}
		}

		// return true if map becomes empty
		return freq.isEmpty();
	}

	public static void main(String[] args) {
		String X = "tommarvoloriddle";	// Tom Marvolo Riddle
		String Y = "iamlordvoldemort";	// I am Lord Voldemort

		if (isAnagram(X, Y)) {
			System.out.println("Anagram");
		} else {
			System.out.println("Not a Anagram");
		}
	}
}
